[ao][sparsity] make sparsity and PTQ compose

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74848
* #74847
* #74846
* __->__ #74845

Summary: This PR adds support for quantization flow to detect
parametrized modules and match them using their original module types.
This mainly involved using the new type_before_parametrizations function rather than
type to check for module mathcing

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240274](https://our.internmc.facebook.com/intern/diff/D35240274)

HDCharles committed:
Make quantization work with sparsity parametrizations

Summary: This PR adds support for quantization flow to detect
parametrized modules and match them using their original module types

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "Make quantization work with sparsity parametrizations"


Summary: This PR adds support for quantization flow to detect
parametrized modules and match them using their original module types

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] make sparsity and PTQ compose"


Summary: This PR adds support for quantization flow to detect
parametrized modules and match them using their original module types.
This mainly involved using a new nonparam_type function rather than
type to check for module mathcing

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] make sparsity and PTQ compose"


Summary: This PR adds support for quantization flow to detect
parametrized modules and match them using their original module types.
This mainly involved using a new nonparam_type function rather than
type to check for module mathcing

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] make sparsity and PTQ compose"


Summary: This PR adds support for quantization flow to detect
parametrized modules and match them using their original module types.
This mainly involved using a new nonparam_type function rather than
type to check for module mathcing

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] make sparsity and PTQ compose"


Summary: This PR adds support for quantization flow to detect
parametrized modules and match them using their original module types.
This mainly involved using a new nonparam_type function rather than
type to check for module mathcing

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] make sparsity and PTQ compose"


Summary: This PR adds support for quantization flow to detect
parametrized modules and match them using their original module types.
This mainly involved using a new nonparam_type function rather than
type to check for module mathcing

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/utils.py:
```diff
@@ -360,3 +361,26 @@ def _parent_name(target):
     else:
         return r[0], r[1]
+
+def nonparam_type(module):
```
can we call it something like `type_before_parametrizations`?

HDCharles replied:
sure

vkuzo commented on torch/ao/quantization/utils.py:
```diff
@@ -360,3 +361,26 @@ def _parent_name(target):
+    else:
+        return type(module)
+
+def is_leaf(module):
```
`is_leaf` is already used in the quantization codebase for the FX use case (a leaf is a module not traced through), could we pick a different name for this?

HDCharles replied:
yeah, I can do something like parametrized_check_for_leaf or something like that?

HDCharles replied:
i changed it to is_leaf_or_only_parametrized

vkuzo replied:
how about something like `has_no_children_ignoring_parametrizations`?

HDCharles replied:
ok

HDCharles committed:
Update on "[ao][sparsity] make sparsity and PTQ compose"


Summary: This PR adds support for quantization flow to detect
parametrized modules and match them using their original module types.
This mainly involved using a new nonparam_type function rather than
type to check for module mathcing

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] make sparsity and PTQ compose"


Summary: This PR adds support for quantization flow to detect
parametrized modules and match them using their original module types.
This mainly involved using the new type_before_parametrizations function rather than
type to check for module mathcing

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][sparsity] make sparsity and PTQ compose"


Summary: This PR adds support for quantization flow to detect
parametrized modules and match them using their original module types.
This mainly involved using the new type_before_parametrizations function rather than
type to check for module mathcing

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240274).

HDCharles committed:
Update on "[ao][sparsity] make sparsity and PTQ compose"


Summary: This PR adds support for quantization flow to detect
parametrized modules and match them using their original module types.
This mainly involved using the new type_before_parametrizations function rather than
type to check for module mathcing

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240274](https://our.internmc.facebook.com/intern/diff/D35240274)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240274).

HDCharles committed:
Update on "[ao][sparsity] make sparsity and PTQ compose"


Summary: This PR adds support for quantization flow to detect
parametrized modules and match them using their original module types.
This mainly involved using the new type_before_parametrizations function rather than
type to check for module mathcing

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240274](https://our.internmc.facebook.com/intern/diff/D35240274)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240274).

HDCharles committed:
Update on "[ao][sparsity] make sparsity and PTQ compose"


Summary: This PR adds support for quantization flow to detect
parametrized modules and match them using their original module types.
This mainly involved using the new type_before_parametrizations function rather than
type to check for module mathcing

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240274](https://our.internmc.facebook.com/intern/diff/D35240274)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240274).

jerryzh168 commented on test/ao/sparsity/test_composability.py:
```diff
@@ -0,0 +1,94 @@
+        self.assertTrue(isinstance(mod[5], torch.nn.quantized.Linear))
+        self.assertEqual(mod(torch.randn(1, 4, 4, 4)).shape, torch.Size([1, 4, 4, 4]))
+
+    def test_s_prep_before_q_prep(self):
```
nit: maybe we can add a bit more context to this test in docstring about what this is testing

HDCharles committed:
Update on "[ao][sparsity] make sparsity and PTQ compose"


Summary: This PR adds support for quantization flow to detect
parametrized modules and match them using their original module types.
This mainly involved using the new type_before_parametrizations function rather than
type to check for module mathcing

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240274](https://our.internmc.facebook.com/intern/diff/D35240274)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240274).

HDCharles committed:
Update on "[ao][sparsity] make sparsity and PTQ compose"


Summary: This PR adds support for quantization flow to detect
parametrized modules and match them using their original module types.
This mainly involved using the new type_before_parametrizations function rather than
type to check for module mathcing

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240274](https://our.internmc.facebook.com/intern/diff/D35240274)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240274).

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240274).
